src/main/java/org/opensearch/security/action/apitokens/ApiToken.java

```diff
@@ -0,0 +1,214 @@
+        List<RoleV7.Index> indexPermissions,
+        Instant creationTime
+    ) {
+        this.description = description;
+        this.jti = jti;
+        this.clusterPermissions = clusterPermissions;
+        this.indexPermissions = indexPermissions;
+        this.creationTime = creationTime;
+
+    }
+
+    public static ApiToken fromXContent(XContentParser parser) throws IOException {
+        String description = null;
+        String jti = null;
+        List<String> clusterPermissions = new ArrayList<>();
+        List<RoleV7.Index> indexPermissions = new ArrayList<>();
+        Instant creationTime = null;
+
+        XContentParser.Token token;
+        String currentFieldName = null;
+
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                switch (currentFieldName) {
+                    case "description":
+                        description = parser.text();
+                        break;
+                    case "jti":
+                        jti = parser.text();
+                        break;
+                    case "creation_time":
+                        creationTime = Instant.ofEpochMilli(parser.longValue());
+                        break;
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                switch (currentFieldName) {
+                    case "cluster_permissions":
+                        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                            clusterPermissions.add(parser.text());
+                        }
+                        break;
+                    case "index_permissions":
+                        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                            if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+                                indexPermissions.add(parseIndexPermission(parser));
+                            }
+                        }
+                        break;
+                }
+            }
+        }
+
+        // Validate required fields
+        if (description == null) {
+            throw new IllegalArgumentException("description is required");
+        }
+        if (jti == null) {
+            throw new IllegalArgumentException("jti is required");
+        }
+        if (creationTime == null) {
+            throw new IllegalArgumentException("creation_time is required");
+        }
+
+        return new ApiToken(description, jti, clusterPermissions, indexPermissions, creationTime);
+    }
+
+    private static RoleV7.Index parseIndexPermission(XContentParser parser) throws IOException {
+        List<String> indexPatterns = new ArrayList<>();
+        List<String> allowedActions = new ArrayList<>();
+        String dls = "";
```

Comment:
Supporting FLS/DLS/Field Masking will introduce complications on token issuance. As part of a check to see that a user can issue a token it needs to be 2 criteria:

1. The user has permissions to issue a token
2. The permissions associated with the token are a subset of the permissions that the user has. With Fls/Dls/Field Masking its not clear how to achieve #2.

I think we should make tokens simple from the start and only allow index patterns and allowed_actions for index permissions.

Reply:
This is a good callout. I made a sub - IndexPermission class within ApiToken which represents the barebones index_pattern and allowed_actions only, will dis-regard the dls, fls, and field masking for now.